### ISSUE TITLE
Navmesh doesn't always update when obstacles change

### ISSUE THREAD

**Issue by Shatur** [vleue/vleue_navigator#55]
I recently tried switching from oxidized_navigation to vleue_navigator. However, I noticed that the navmesh doesn't update, and #54 didn't fix the issue. So, I switched back to oxidized_navigation in [this](https://github.com/projectharmonia/project_harmonia/commit/86ba69d7883a33f7295bc1def1af620270aa15c2) commit.

@mockersf asked me on Discord if he could try running the project before the commit. Since I’ve fixed quite a few bugs since then, I decided to create a separate branch with this commit reverted and rebased onto the latest master: #330. Hopefully, this will help determine whether it's a bug or an API misuse on my part.

Below are the steps to reproduce. It requires a bit of setup when you try it for the first time :)

1. `cargo run` if you running it for the first time. Next time run `cargo run -- play --world-name "New world" family "New family"` instead and jump directly to step 16. You may also want to run with `RUST_LOG=project_harmonia_base=debug` to tracing.
2. Click on "Play":
![Screenshot_20240919_022120](https://github.com/user-attachments/assets/c5755754-46cf-4c39-a0bb-3e9f42fd9f9f)
3. First we need to create a world. Just click on "Create":
![image](https://github.com/user-attachments/assets/4dfad89e-f27e-4423-adb8-711504c75bd9)
4. Confirm creation, just leave the default name:
![image](https://github.com/user-attachments/assets/585b3fa7-eb25-4b72-b717-5d43f987a925)
5. Now we need to create a city first. Not very intuitive, but select cities tab first:
![image](https://github.com/user-attachments/assets/c430e83e-6c44-4192-9041-91975481acc5)
6. Press "Create":
![image](https://github.com/user-attachments/assets/0381fff8-3ecb-45a0-9783-aa1ce0ab0e92)
7. Confirm creation, just leave the default name:
![image](https://github.com/user-attachments/assets/8b7f0a10-3e73-4fce-9542-caa938c509d8)
8. Now switch back to the families tab:
![image](https://github.com/user-attachments/assets/0cadbe32-598d-4c52-8dae-33fa7357d695)
9. Click on "Create new":
![image](https://github.com/user-attachments/assets/4f4edbc8-a5d6-4d28-acf7-ab0f279441c2)
10. It's a stub for character creation screen :sweat_smile: Just click on "Confirm":
![image](https://github.com/user-attachments/assets/6feff588-4487-4ca8-9912-d0fd694f89c5)
11. Confirm the family name:
![image](https://github.com/user-attachments/assets/57d76701-aa3a-4fcc-9f99-bf66dbcc09ec)
12. Now click on "Place & save":
![image](https://github.com/user-attachments/assets/cf18c831-b30d-43f6-aa95-112c877a1a5b)
13. Press Esc and click on "Save". It's very important, if you exit the game the world won't be saved!
![image](https://github.com/user-attachments/assets/7200b916-559a-4666-a22a-9f1bea24c311)
14. (Optional) You probably want to see navigation paths and navmesh. Let's enable them. Press Esc and click on "Settings":
![image](https://github.com/user-attachments/assets/9ccc6e6d-8abf-4054-90f4-9f9ea8cf7a61)
15. (Optional) Navigate to developer tab:
![image](https://github.com/user-attachments/assets/e37a8cdc-5ff0-4369-ab90-fdcccb726939)
16. (Optinal) Tick the needed checkboxes and press "Ok":
![image](https://github.com/user-attachments/assets/9788fdf5-7775-448c-967f-62d4100bbd13)
17. Now in order to place walls you need to enter build mode:
![image](https://github.com/user-attachments/assets/db97520a-79b4-4a25-a482-5b837ee04266)
18. Select walls:
![image](https://github.com/user-attachments/assets/7b0aec4e-d0a7-403f-afdb-a96370b70fc7)
19. Click anywhere to start placing a wall:
![image](https://github.com/user-attachments/assets/2c630b78-063b-4120-bdcd-0a6e94d68739)
20. Click anywhere again to confirm placing. You can already see that the navmesh is not updated. Let's try to move a character, back to the family mode:
![image](https://github.com/user-attachments/assets/7243057b-ce56-4021-bf9f-8e0204e1d496)
21. Click somewhere behind the wall and select "Walk here".
![image](https://github.com/user-attachments/assets/009200b0-789b-440f-8f5f-560526cb0ef7)
22. You can see how the character goes through the wall:
![image](https://github.com/user-attachments/assets/b6239bbf-b37d-4479-9e94-bc4ce2020b38)

Since I have networking and serialization, I use the [blueprint pattern](https://docs.rs/bevy_replicon/latest/bevy_replicon/#blueprints-pattern), which is basically "required components at home". For walls you can see that an empty collider is inserted [here](https://github.com/projectharmonia/project_harmonia/blob/bf06e77e7ff17f9c88e431e72a20993f8b34cdff/base/src/game_world/family/building/wall.rs#L74). And collision is updated [here](https://github.com/projectharmonia/project_harmonia/blob/bf06e77e7ff17f9c88e431e72a20993f8b34cdff/base/src/game_world/family/building/wall.rs#L120) on the wall change.
Let me know any questions :)

After writing this I started to doubt that it's the best way to debug the issue :laughing: Maybe it's simpler to just create a simple example with a changing mesh.

**Comment by mockersf** [comment]
Does #56 fixes it for you?

On my local it works and navmesh updates are instant

https://github.com/user-attachments/assets/7b76de97-2d4c-4e0d-824b-ea027b04de19

**Comment by Shatur** [comment]
Ah, yes, setting `upward_shift` to 0.1 fixes the issue, thank you! I tried it before, but with negative values :sweat_smile: Looks like navmesh now always updates, we can close this issue.

But I faced another issue, I trying to run the world from [this](https://mastodon.social/@Shatur/113160041777355607) post, but character is unable to walk through doors with weird errors:
```
2024-09-20T07:53:38.122797Z ERROR polyanya: Search from (-24.823553, -15.517356) to (-18.67345, -31.258448) failed. Please check the mesh is valid as this should not happen. 
```
When a door is placed on a wall, it splits the collider using width from the placed object:
![image](https://github.com/user-attachments/assets/8353adde-4f36-4aaa-b35c-585d7e5542e2)
Ignore the door collision, it's a bit broken, but it doesn't affect navigation, only walls are marked as `Obstacle`.

Here is the scene: 
[New world.zip](https://github.com/user-attachments/files/17072006/New.world.zip)
The game uses standard paths to store data, so on Linux it's `~/.config/project_harmonia_base/worlds/` (I use crate name, but recently refactored into multiple crates, need to fix it :smile:).
